cmd/lotus-chainwatch/processor/processor.go

```diff
@@ -0,0 +1,309 @@
+
+func (p *Processor) refreshViews() error {
+	if _, err := p.db.Exec(`refresh materialized view state_heights`); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Processor) collectActorChanges(ctx context.Context, toProcess map[cid.Cid]*types.BlockHeader) (map[cid.Cid]ActorTips, error) {
+	start := time.Now()
+	defer func() {
+		log.Infow("Collected Actor Changes", "duration", time.Since(start).String())
+	}()
+	// ActorCode - > tipset->[]actorInfo
+	out := map[cid.Cid]ActorTips{}
+	var outMu sync.Mutex
+
+	// map of addresses to changed actors
+	var changes map[string]types.Actor
+	actorsSeen := map[cid.Cid]struct{}{}
+
+	// collect all actor state that has changes between block headers
+	paDone := 0
+	parmap.Par(50, parmap.MapArr(toProcess), func(bh *types.BlockHeader) {
+		paDone++
+		if paDone%100 == 0 {
+			log.Infow("Collecting actor changes", "done", paDone, "percent", (paDone*100)/len(toProcess))
+		}
+
+		pts, err := p.node.ChainGetTipSet(ctx, types.NewTipSetKey(bh.Parents...))
+		if err != nil {
+			panic(err)
+		}
+
+		// collect all actors that had state changes between the blockheader parent-state and its grandparent-state.
+		// TODO: changes will contain deleted actors, this causes needless processing further down the pipeline, consider
+		// a separate strategy for deleted actors
+		changes, err = p.node.StateChangedActors(ctx, pts.ParentState(), bh.ParentStateRoot)
+		if err != nil {
+			panic(err)
+		}
+
+		// record the state of all actors that have changed
+		for a, act := range changes {
+			act := act
+			a := a
+
+			addr, err := address.NewFromString(a)
+			if err != nil {
+				panic(err)
+			}
+
+			ast, err := p.node.StateReadState(ctx, addr, pts.Key())
+			if err != nil {
+				panic(err)
+			}
+
+			// TODO look here for an empty state, maybe thats a sign the actor was deleted?
+
+			state, err := json.Marshal(ast.State)
+			if err != nil {
+				panic(err)
+			}
+
+			outMu.Lock()
+			if _, ok := actorsSeen[act.Head]; !ok {
+				_, ok := out[act.Code]
+				if !ok {
+					out[act.Code] = map[types.TipSetKey][]actorInfo{}
+				}
+				out[act.Code][pts.Key()] = append(out[act.Code][pts.Key()], actorInfo{
+					act:         act,
+					stateroot:   bh.ParentStateRoot,
+					height:      bh.Height,
+					tsKey:       pts.Key(),
+					parentTsKey: pts.Parents(),
+					addr:        addr,
+					state:       string(state),
+				})
+			}
+			actorsSeen[act.Head] = struct{}{}
+			outMu.Unlock()
+		}
+	})
+	return out, nil
+}
+
+func (p *Processor) unprocessedBlocks(ctx context.Context, batch int) (map[cid.Cid]*types.BlockHeader, error) {
+	start := time.Now()
+	defer func() {
+		log.Infow("Gathered Blocks to process", "duration", time.Since(start).String())
+	}()
+	rows, err := p.db.Query(`
+with toProcess as (
+    select blocks.cid, blocks.height, rank() over (order by height) as rnk
+    from blocks
+        left join blocks_synced bs on blocks.cid = bs.cid
+    where bs.is_processed is false and blocks.height != 0
```

Comment:
A slightly better condition which excludes the possibility of negative ints and NULLs.
```suggestion
    where bs.is_processed is false and blocks.height > 0
```

Reply:
Also, I saw you mentioned `processed_at` and `is_processed` are both included in this table. Consider using only `processed_at` which can be NULL when is_processed = false.